src/apple/macos/component/arm.rs

```diff
@@ -0,0 +1,175 @@
+// Take a look at the license at the top of the repository in the LICENSE file.
+
+use std::ffi::CStr;
+
+use core_foundation_sys::array::{CFArrayGetCount, CFArrayGetValueAtIndex};
+use core_foundation_sys::base::kCFAllocatorDefault;
+use core_foundation_sys::string::{
+    kCFStringEncodingUTF8, CFStringCreateWithBytes, CFStringGetCStringPtr,
+};
+
+use super::super::CFReleaser;
+use crate::apple::inner::ffi::{
+    kHIDPage_AppleVendor, kHIDUsage_AppleVendor_TemperatureSensor, kIOHIDEventTypeTemperature,
+    matching, IOHIDEventFieldBase, IOHIDEventGetFloatValue, IOHIDEventSystemClientCopyServices,
+    IOHIDEventSystemClientCreate, IOHIDEventSystemClientSetMatching, IOHIDServiceClientCopyEvent,
+    IOHIDServiceClientCopyProperty, __IOHIDEventSystemClient, __IOHIDServiceClient,
+    HID_DEVICE_PROPERTY_PRODUCT,
+};
+use crate::ComponentExt;
+
+pub(crate) struct Components {
+    pub inner: Vec<Component>,
+    client: Option<CFReleaser<__IOHIDEventSystemClient>>,
+}
+
+impl Components {
+    pub(crate) fn new() -> Self {
+        Self {
+            inner: vec![],
+            client: None,
+        }
+    }
+
+    pub(crate) fn refresh(&mut self) {
+        self.inner.clear();
+
+        let client = || -> Option<CFReleaser<_>> {
+            unsafe {
+                let matches = CFReleaser::new(matching(
+                    kHIDPage_AppleVendor,
+                    kHIDUsage_AppleVendor_TemperatureSensor,
+                ))?;
+
+                let client = CFReleaser::new(IOHIDEventSystemClientCreate(kCFAllocatorDefault))?;
+
+                let _ = IOHIDEventSystemClientSetMatching(client.inner(), matches.inner());
+                Some(client)
+            }
+        }();
+
+        if client.is_none() {
+            return;
+        }
+        let client = client.unwrap();
+
+        unsafe {
+            let services = IOHIDEventSystemClientCopyServices(client.inner());
```

Comment:
`services` needs to be freed using `CFRelease`. And before every `return` too. The best thing you can do is wrapping it into a struct which implements `Drop` (which calls `CFRelease` automatically).

Reply:
Ah you already added `CFReleaser` so you already have the needed struct. :)